File: CTDL & GT/stack/stack.cpp
/*
    Basic Stack C++    
*/

#include <iostream>
#include <time.h>

using namespace std;

template<typename T>
struct Stack{
    T* arr;
    int length;
    int capacity;

    Stack() : length(0), capacity(10){ arr = new T[capacity];}
    ~Stack() { delete[] arr; }

    T& operator[](T index){
        if (index < 0 || index >= length) 
            throw out_of_range("Index out of range");
        
        return arr[index];
    }

    void resize(){
        int newSpace = capacity*2;
        T* tmp = new T[newSpace];
        for (int i = 0; i < length; i++)
            tmp[i] = arr[i];

        delete[] arr;
        arr = tmp;
        capacity = newSpace;
    }
    void push(T value){
        if(arr == nullptr){
            arr = new T[1];
            arr[0] = value;
            length++;
            return;
        }
        if(length == capacity)
            resize();

        arr[length++] = value;
    }

    T pop(){
        if(arr == nullptr){
            cout<<"Stack is an Empty\n";
            return T();
        }
        T res = arr[length-1];
        
        T* tmp = new T[length-1];
        for (int i = 0; i < length-1; i++)
            tmp[i] = arr[i];
        
        delete[] arr;
        arr = tmp;
        length--;

        return res;
    }

    T top(){
        if(arr == nullptr){
            cout<<"Stack is an Empty\n";
            return T();
        }   
        return arr[length-1];
    }

    bool isEmpty(){
        return length==0;
    }
};

int main(){
    srand(time(nullptr));
    Stack<int> s;
    for(int i = 1; i <= 10; i++)
        s.push(rand() % 100 + 1);

    cout<<(s.isEmpty() ? "Is empty\n" : "Is available\n");
    for (int i = 0; i < s.length; i++)
        cout<<s[i]<<' ';

    s.pop();
    cout<< "\nTop element after pop: "<< s.top();

    cout<<"\nlength: "<<s.length;
    cout<<"\ncapacity: "<<s.capacity;
    cout<<endl;

    return 0;
    
}

File: CTDL & GT/stack/stack.rs
/*
    Stack Rust
*/

use std::time::SystemTime;

struct Stack<T> {
    arr: Vec<T>,
    length: usize,
    capacity: usize,
}

impl<T> Stack<T> {
    fn new() -> Self {
        Stack {
            arr: Vec::with_capacity(10),
            length: 0,
            capacity: 10,
        }
    }

    fn resize(&mut self) {
        self.capacity *= 2;
        self.arr.reserve(self.capacity - self.arr.len());
    }

    fn push(&mut self, value: T) {
        if self.length == self.capacity {
            self.resize();
        }
        self.arr.push(value);
        self.length += 1;
    }

    fn pop(&mut self) -> Option<T> {
        if self.length == 0 {
            println!("Stack is empty");
            None
        } else {
            self.length -= 1;
            self.arr.pop()
        }
    }

    fn top(&self) -> Option<&T> {
        if self.length == 0 {
            println!("Stack is empty");
            None
        } else {
            self.arr.get(self.length - 1)
        }
    }

    fn is_empty(&self) -> bool {
        self.length == 0
    }
}

fn main() {
    let mut s = Stack::new();
    let mut rng = rand::thread_rng();
    for _ in 1..=10 {
        s.push(rand::random::<u32>() % 100 + 1);
    }

    println!("{}", if s.is_empty() { "Is empty" } else { "Is available" });
    for i in 0..s.length {
        println!("{}", s.arr[i]);
    }

    s.pop();
    if let Some(top) = s.top() {
        println!("\nTop element after pop: {}", top);
    }

    println!("\nlength: {}", s.length);
    println!("\ncapacity: {}", s.capacity);
}
